zerver/tests/test_decorators.py

```diff
@@ -647,58 +647,62 @@ class TestLoggingErrorException(Exception):
 
         return mock.patch("logging.error", side_effect=TestLoggingErrorException)
 
-    def check_rate_limit_public_or_user_views(self, remote_addr: str, client_name: str) -> None:
-        META = {"REMOTE_ADDR": remote_addr}
-
-        request = HostRequestMock(client_name=client_name, meta_data=META)
+    def check_rate_limit_public_or_user_views(
+        self, remote_addr: str, client_name: str, expect_rate_limit: bool
+    ) -> None:
+        META = {"REMOTE_ADDR": remote_addr, "PATH_INFO": "test"}
 
-        f = self.get_ratelimited_view()
+        request = HostRequestMock(host="zulip.testserver", client_name=client_name, meta_data=META)
 
         with mock.patch(
             "zerver.lib.rate_limiter.rate_limit_user"
         ) as rate_limit_user_mock, mock.patch(
             "zerver.lib.rate_limiter.rate_limit_ip"
-        ) as rate_limit_ip_mock:
-            with self.errors_disallowed():
-                self.assertEqual(orjson.loads(f(request).content).get("msg"), "some value")
+        ) as rate_limit_ip_mock, self.errors_disallowed():
+            self.assert_in_success_response(["some value"], self.public_view(request))
+        self.assertEqual(rate_limit_ip_mock.called, expect_rate_limit)
+        self.assertFalse(rate_limit_user_mock.called)
 
+        # We need to recreate the request, because process_client mutates client and client_name on
+        # the associated RequestNotes, causing the request to be incorrectly rate limited.
```

Comment:
The comment might be okay the way it is, but for my understanding can you explain what does `process_client` do that causes incorrect rate limiting? :thinking:

Reply:
`should_rate_limit` check the client name to determine if the request should be rate limited; in this case, `process_client` changes the client on the `RequestNotes`. Reading this again, I realized that only `client` but not `client_name` get mutated. Will update the comment.

Reply:
Ah, so basically this is relevant for test cases involving `client_is_exempt_from_rate_limiting`?

Reply:
Yes.

Reply:
Although it technically can also cause false negatives when `client_is_exempt_from_rate_limiting` is expected to be `False`.